File: 프로그래머스/LV2/연속된 부분 수열의 합.py
def solution(sequence, k):
    answer = []
    start, end = 0, 0
    total = sequence[0]

    while True:
        if total == k:
            if len(answer) == 0:
                answer.append(start)
                answer.append(end)
            elif answer[1] - answer[0] > end - start:
                answer.pop()
                answer.pop()
                answer.append(start)
                answer.append(end)
            total -= sequence[start]
            start += 1

        elif total > k:
            total -= sequence[start]
            start += 1
        else:
            end += 1
            if end == len(sequence):
                break
            total += sequence[end]

    return answer

File: 프로그래머스/javascript/LV2/연속된 부분 수열의 합.js
// 투 포인터 이용
function solution(sequence, k) {
  let left = 0;
  let right = 0;
  let sum = sequence[0];
  let answer = [0, sequence.length - 1];

  while (left <= right && right < sequence.length) {
    if (sum === k) {
      if (right - left < answer[1] - answer[0]) {
        answer = [left, right];
      }
      sum -= sequence[left];
      left++;
    } else if (sum < k) {
      right++;
      sum += sequence[right];
    } else {
      sum -= sequence[left];
      left++;
    }
  }

  return answer;
}
